### ISSUE TITLE
Controller input to patterns

### ISSUE THREAD

**Issue by yaxu** [tidalcycles/Tidal#344]
Currently patterns are pure functions of time, something like `Arc -> [((Arc,Arc),Value)]`. 

Up 'til now we've called the patterns of synth params `ParamPattern`s. I propose renaming them to `ControlPattern`s.

In particular
```
type ControlMap = Map String Value
type ControlPattern = Pattern ControlMap
```

So how about if patterns were functions not only of time, but of control state? Button presses, knob twiddles, global temperature, etc.. Then a Pattern could be redefined as taking state as input, including time but also possibly many other values:

```
data Nature = Analog | Digital
type State = (Arc, ControlMap)
data Pattern a = Pattern (query :: State -> [((Arc,Arc),a)], nature :: Nature)
```

On the face of it, this would be straightforward to implement, allowing for example:
```
sound "arpy" >| n "0 ~ 3 2" |> speed ("2 3" + control "dial1")
```
Where `control` simply takes the value from the `ControlMap` inside the `State` that is passed to the pattern.

**Comment by yaxu** [comment]
The big problem with this would be latency. We could try rendering patterns at a steady e.g. 100Hz rather than the current default (I think) 8Hz. We could try lowering the tidal > superdirt latency too. I think it's still going to feel sluggish, though.

An alternative could be for tidal to not deal with state itself, but compose controls and values into a little algebra, which could then be sent to superdirt, for lower latency processing in sclang or scsynth..

**Comment by yaxu** [comment]
Perhaps you have some thoughts from a superdirt perspective, @telephon?

**Comment by yaxu** [comment]
Seems to work, I have `controlI`, `controlS` and `controlF` for integer, string and float controls respectively.
It'd be easy enough to open up an osc port and receive messages in the form `/control flurble 0.4`, which can then be accessed by a pattern with `sound "bd sn" # speed (controlI "flurble")`

**Comment by yaxu** [comment]
I guess the next question is whether events should be able to modify the state..

**Comment by cannc4** [comment]
How about modifying the state from another pattern? e.g using another pattern as a controller

**Comment by cannc4** [comment]
something like this would sound great:
`d1 $ sound "bd sn" # speed (controlI "flurble")` 
then
`d2 $ n "0.4 [1.4 0.8 ]" #s "flurble" `

**Comment by yaxu** [comment]
I was thinking along similar lines @cannc4, if controls come from OSC, and patterns send OSC, then it could be plugged together quite easily.. Or could engineer a more direct way of patterning controls.

`d2 $ flurble "0.4 [1.4 0.8 ]"` would be doable

**Comment by cannc4** [comment]
That'd be pretty nice.

**Comment by telephon** [comment]
>Perhaps you have some thoughts from a superdirt perspective, @telephon?

We could easily offload this feature to superdirt, e.g. by interpreting symbols instead of numbers as variable names that refer to either sclang state or to scsynth busses. 

The main trade off this the following:

1. pros: that any state change happens exactly when the event sounds, and may even be continuous in change while the event plays.

2. cons: any calculation on state changes have to be specified in sclang or ugens, and calculations in tidal do not apply to the values of these states.

In general, this is a typical case where the lack of integration is a disadvantage (we have the same actually between sclang and scsynth). To a degree, this could be alleviated, if operations on symbols in tidal would return a representation of calculations that sclang could interpret as a function to compute. But in the final limit, this would  mean to construct ghci in sclang! Or, alternatively, it would mean to implement a graph representation of all tidal functions that can be translated in a graph of synths for scserver. This would then be a good first step to get rid of superdirt and have it all in tidal! :)

**Comment by yaxu** [comment]
If we had ASTs everywhere (#217), I think it would be doable to send that third event as a string containing `"ctl1+0.5"`. We don't at the moment, though.. and true, if you did want to apply a more complex function to a control you'd hit problems.

It's turning out to be easy to do this on the tidal side though, just by passing state in with time. I think it's probably best to focus on this side, and trying to get latency right down, in a configurable way.

**Comment by telephon** [comment]
yes. Still it would also be something to be able to have properly continuous controls. We could have both, first the tidal side one for calculation flexibility, and a really simple one on superdirt, where you get continuous control on demand? For that we'd have to allow strings to be sent in place of floats for all ControlPatterns.

**Comment by telephon** [comment]
Sorry, I haven't understood this:
>that sends the same name but a different type.

**Comment by yaxu** [comment]
sorry was quite tired yesterday and deleted most of what I wrote because I was having problems articulating it..
But would be straightforward in the refactor branch to sometimes send floats and sometimes strings for the same parameter name.
I was also thinking though that tidal could also send e.g. `speed_mul` and `speed_add` parameters, to be multiplied or added together in superdirt. Then that would be a simple way to e.g. scale external controller state values with tidal patterns.

**Comment by telephon** [comment]
Yes, that could make sense. I wouldn't want to parse every parameter string and search for a "_mul", but if these are fixed, it would work. Now the most interesting place to have those is actually where you could do continuous mapping. Everything else is probably better done in tidal.

If I look at the parameters of the core sample player, then these could just have an extra set of controls each for mul and add for these:

`sustain = 1, begin = 0, end = 1, speed = 1, endSpeed = 1, freq = 440, pan = 0`

And it would be up to anyone who liked to add more or others.

But I hesitate a little bit because the decision which parameters to support and which not is pretty arbitrary. I can think of some very dirty secret way to sneak in the implementations into SynthDefs, though.

**Comment by cannc4** [comment]
> I can think of some very dirty secret way to sneak in the implementations into SynthDefs, though.

Could perhaps use buffers to store parameters and read them within the Synthdefs

**Comment by telephon** [comment]
no, that is not needed, we can map controls to busses. The implementation is easy, it is just harder to make it general enough to control from tidal.

**Comment by msp** [comment]
> But would be straightforward in the refactor branch to sometimes send floats and sometimes strings for the same parameter name.

Wouldn't that be confusing if you're trying to use Tidal to pattern anything other than Superdirt? Perhaps that's not a common use case..
